src/impl_ptrs.rs

```diff
@@ -0,0 +1,172 @@
+use dimension::stride_offset;
+use imp_prelude::*;
+use StrideShape;
+
+impl<A, D> ArrayPtr<A, D>
+where
+    D: Dimension,
+{
+    /// Create a new `ArrayPtr`.
+    ///
+    /// Unsafe because caller is responsible for ensuring that pointers to all
+    /// elements in the array are within bounds or one byte past the end of a
+    /// **single** allocated object.
+    #[inline(always)]
+    pub(crate) unsafe fn new_(ptr: *const A, dim: D, strides: D) -> Self {
+        ArrayPtr {
+            data: PtrRepr::new(),
+            ptr: ptr as *mut A,
+            dim: dim,
+            strides: strides,
+        }
+    }
+
+    /// Create an `ArrayPtr<A, D>` from shape information and a raw pointer to
+    /// the elements.
+    ///
+    /// Unsafe because caller is responsible for ensuring that pointers to all
+    /// elements in the array are within bounds or one byte past the end of a
+    /// **single** allocated object.
+    pub unsafe fn from_shape_ptr<Sh>(shape: Sh, ptr: *const A) -> Self
+    where
+        Sh: Into<StrideShape<D>>,
+    {
+        let shape = shape.into();
+        let dim = shape.dim;
+        let strides = shape.strides;
+        ArrayPtr::new_(ptr, dim, strides)
+    }
+
+    /// Return a read-only view of the array.
+    ///
+    /// **Warning** from a safety standpoint, this is equivalent to
+    /// dereferencing a raw pointer for every element in the array. You must
+    /// ensure that all of the data is valid and choose the correct lifetime.
+    #[inline]
+    pub unsafe fn deref_view<'a>(&self) -> ArrayView<'a, A, D> {
+        ArrayView::new_(self.ptr, self.dim.clone(), self.strides.clone())
+    }
+
+    /// Converts to a read-only view of the array.
+    ///
+    /// **Warning** from a safety standpoint, this is equivalent to
+    /// dereferencing a raw pointer for every element in the array. You must
+    /// ensure that all of the data is valid and choose the correct lifetime.
+    #[inline]
+    pub(crate) unsafe fn deref_into_view<'a>(self) -> ArrayView<'a, A, D> {
+        ArrayView::new_(self.ptr, self.dim, self.strides)
+    }
```

Comment:
This into method seems to be the most interesting to provide, so I'd go with it to be pub?

Reply:
Okay, we can make `deref_into_view` public. Would you remove `deref_view`?

Reply:
Yes I'd consider removing it

Reply:
Okay, done. The user can use `raw_view.clone().deref_into_view()` if they don't want to give up ownership of the raw view. We can add `.deref_view()` later if it turns out that it would be useful.